sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1160 @@
+# coding: utf-8
+"""
+Large Margin Nearest Neighbor Classification
+"""
+
+# Author: John Chiotellis <johnyc.code@gmail.com>
+# License: BSD 3 clause
+
+from __future__ import print_function
+from warnings import warn
+
+import sys
+import time
+import numpy as np
+from scipy.optimize import minimize
+from scipy.sparse import csr_matrix, csc_matrix, coo_matrix
+from six import integer_types, string_types
+
+from ..base import BaseEstimator, TransformerMixin
+from ..neighbors import NearestNeighbors
+from ..decomposition import PCA
+from ..utils import gen_batches
+from ..utils.extmath import row_norms, safe_sparse_dot
+from ..utils.extmath import _euclidean_distances_without_checks
+from ..utils.random import check_random_state
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import (check_is_fitted, check_array, check_X_y,
+                                check_scalar)
+from ..exceptions import ConvergenceWarning
+
+
+class LargeMarginNearestNeighbor(BaseEstimator, TransformerMixin):
+    """Distance metric learning for large margin classification.
+
+    Read more in the :ref:`User Guide <lmnn>`.
+
+    Parameters
+    ----------
+    n_neighbors : int, optional (default=3)
+        Number of neighbors to use as target neighbors for each of the samples.
+
+    n_components : int, optional (default=None)
+        Preferred dimensionality of the embedding.
```

Comment:
I'm hesitating to call this a "projection", since it's not enforced in any way, as it is for example in eigendecomposition-based methods. If anyone else has a strong opinion on this, please comment.

Reply:
I just looked up the definition of projection, and indeed, strictly speaking LMNN is not a projection since a projection should be idempotent. But neither `PCA`, or `RandomProjection` for instance are projections according to this definition...
But then maybe "transformation" would be a better term ? It's just that such a term allows the users to quickly understand that they will be able to transform new points with a transformation matrix, which is not the case for a lot of manifold learning algorithms

Reply:
Yes, see my other comment. Strictly speaking, (linear) transformation or mapping is the right term